include/dali/core/fast_div.h

```diff
@@ -0,0 +1,247 @@
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef DALI_CORE_FAST_DIV_H_
+#define DALI_CORE_FAST_DIV_H_
+
+#include <stdio.h>
+#include <cuda_runtime.h>
+#include <cstdint>
+#include "dali/core/util.h"
+#include "dali/core/host_dev.h"
+
+namespace dali {
+
+namespace detail {
+
+template <typename component>
+struct lohi {
+  component lo, hi;
+};
+
+template <typename T>
+DALI_HOST_DEV DALI_FORCEINLINE lohi<T> operator<<(lohi<T> x, unsigned sh) {
+  using U = typename std::make_unsigned<T>::type;
+  static constexpr unsigned bits = sizeof(T) * 8;
+  if (sh == 0) {
+    return x;
+  } else if (sh >= bits) {
+    return { 0, x.lo << (sh-bits) };
+  } else {
+    return {
+      x.lo << sh,
+      x.hi << sh | U(x.lo) >> (bits - sh)
+    };
+  }
+}
+
+template <typename T>
+DALI_HOST_DEV DALI_FORCEINLINE lohi<T> operator>>(lohi<T> x, unsigned sh) {
+  using U = typename std::make_unsigned<T>::type;
+  static constexpr unsigned bits = sizeof(T) * 8;
+  if (sh == 0) {
+    return x;
+  } else if (sh >= bits) {
+    return { x.hi >> (sh-bits), 0 };
+  } else {
+    return {
+      U(x.lo) >> sh | x.hi << (bits - sh),
+      x.hi >> sh
+    };
+  }
+}
+
+template <typename T>
+DALI_HOST_DEV DALI_FORCEINLINE lohi<T> &operator<<=(lohi<T> &x, unsigned sh) {
+  x = x << sh;
+  return x;
+}
+
+template <typename T>
+DALI_HOST_DEV DALI_FORCEINLINE lohi<T> &operator>>=(lohi<T> &x, unsigned sh) {
+  x = x >> sh;
+  return x;
+}
+
+template <typename T>
+DALI_HOST_DEV DALI_FORCEINLINE lohi<T> operator-(lohi<T> a, lohi<T> b) {
+  lohi<T> ret;
+  ret.lo = a.lo - b.lo;
+  int borrow = (b.lo && ret.lo > a.lo);
+  ret.hi = a.hi - b.hi - borrow;
+  return ret;
+}
+
+DALI_HOST_DEV inline uint32_t div_lohi(uint32_t lo, uint32_t hi, uint32_t divisor) {
+  return (static_cast<uint64_t>(hi) << 32 | lo) / divisor;
+}
+
+DALI_HOST_DEV DALI_FORCEINLINE lohi<uint64_t> mull(uint64_t a, uint64_t b) {
+  lohi<uint64_t> ret;
+#ifdef __CUDA_ARCH__
+  ret.lo = a * b;
+  ret.hi = __umul64hi(a, b);
+#else
+  unsigned __int128 m = (unsigned __int128)a * b;
+  ret.lo = m;
+  ret.hi = m >> 64;
+#endif
+  return ret;
+}
+
+using dali::ilog2;
+
+template <typename uint>
+DALI_HOST_DEV DALI_FORCEINLINE int ilog2(lohi<uint> lh) {
+  return lh.hi ? ilog2(lh.hi) + sizeof(uint)*8 : ilog2(lh.lo);
+}
+
+DALI_HOST_DEV inline uint64_t div_lohi(uint64_t lo, uint64_t hi, uint64_t divisor) {
+#if defined(__x86_64) && defined(__GNUC__) && !defined(__CUDA_ARCH__)
+  // I hope this gets compiled to dividing rdx:rax register pair by a 64-bit value
```

Comment:
> I hope this 

Brave assumption

Reply:
Unfortunately, doens't work, it emits a call to a library function.  Still, It's likely faster than the long division I implemented below.
x86 cannot do true 128-bit division - if the quotient exceeds register width, division overflow error is raised - but it's actually what I want. Maybe I'll use an Intel intrinsic, if I can find one.

Reply:
No intrinsic (!) - but I believe that the library function will make good use of div instruction in case where no overflow is going to occur.